x/derivatives/keeper/margin.go

```diff
@@ -6,6 +6,94 @@ import (
 	"github.com/UnUniFi/chain/x/derivatives/types"
 )
 
+func (k Keeper) AddMargin(ctx sdk.Context, sender sdk.AccAddress, positionId string, amount sdk.Coin) error {
+	position := k.GetPositionWithId(ctx, positionId)
+	if position == nil {
+		return types.ErrPositionDoesNotExist
+	}
+
+	// Send additional margin by sender to the margin manager module account
+	if err := k.SendMarginToMarginManager(ctx, sender, sdk.NewCoins(amount)); err != nil {
+		return err
+	}
+
+	// Add margin to the position
+	position.RemainingMargin = position.RemainingMargin.Add(amount)
+
+	// Make sure if the position is not under the liquidation condition
+	if position.RemainingMargin.IsNegative() {
+		return types.ErrNegativeMargin
+	}
+
+	// Check if the updated margin is not under liquidation
+	params := k.GetParams(ctx)
+	currentBaseUsdRate, currentQuoteUsdRate, err := k.GetPairUsdPriceFromMarket(ctx, position.Market)
+	if err != nil {
+		return err
+	}
+	quoteTicker := k.GetPoolQuoteTicker(ctx)
+	baseMetricsRate := types.NewMetricsRateType(quoteTicker, position.Market.BaseDenom, currentBaseUsdRate)
+	quoteMetricsRate := types.NewMetricsRateType(quoteTicker, position.Market.QuoteDenom, currentQuoteUsdRate)
+
+	if position.NeedLiquidation(params.PerpetualFutures.MarginMaintenanceRate, baseMetricsRate, quoteMetricsRate) {
```

Comment:
This handling may be unnecessary; adding a margin before ReportLiquidation is fine.

Reply:
Liquidation right after adding margin is bad ux.
And allowing adding margin even if it's under liquidation condition doesn't make sense since we shouldn't allow any our-defined insolvent position.
So, this emits an event to tell it.

Reply:
If MsgReportLiquidation is not performed for a long time, the possibility of protocol loss increases.
I have no problem with the position owner noticing and adding margin on his own.

Reply:
I consulted with @KimuraYu45z , No problem allowing margin additions to positions that are needed for liquidation and have not been liquidated.

Reply:
I've fixed it.